torch/csrc/autograd/record_function.cpp

```diff
@@ -189,14 +193,60 @@ void RecordFunction::processCallbacks() {
 }
 
 RecordFunction::~RecordFunction() {
+  end();
+}
+
+void RecordFunction::end() {
   if (initialized_) {
     for (size_t idx = 0; idx < manager().end_callbacks.size(); ++idx) {
       if (!manager().is_callback_sampled[idx] || run_sampled_) {
         manager().end_callbacks[idx](*this);
       }
     }
+
+    AT_ASSERT(thread_local_func_ == this, name_, ": must be top of stack");
     thread_local_func_ = parent_;
+    initialized_ = false;
   }
 }
 
-}}}
+at::Tensor record_function_enter(const std::string& name) {
+  auto rec = at::guts::make_unique<RecordFunction>();
+  // Only add new scope if profiling is enabled.
+  if (thread_local_func_) {
+    AT_ASSERT(
+        thread_local_func_->name() ==
+        StringView("profiler::record_function_enter"));
+    // RecordFunction requires parent_ to be alive for it's entire lifetime.
+    // Since the currently active RecordFunction will only live for the lifetime
+    // of this op we need to end it early so the new RecordFunction we create is
+    // a direct child of the parent RecordFunction.
+    thread_local_func_->end();
+
+    rec->before(name);
+  }
+  return at::cpp_custom_type_hack::create(std::move(rec), at::TensorOptions());
+}
+
+void record_function_exit(const at::Tensor& handle) {
```

Comment:
just from code organization perspective - move it to separate file (like record_function_ops)? we might move profiler around in the future and it's better not to create circular dependencies

Reply:
Unfortunately these ops need to access thread_local_func_ which is only available in this file. @dzhulgakov  Thoughts on how best to resolve that?

Reply:
@d4l3k Make a declaration in torch/csrc/autograd/record_function.h?

Reply:
Ended up adding a static RecordFunction::current() method to return thread_local_func_. Don't really want to expose direct access to it since it's somewhat fragile.